File: hto/67/hto_67.cpp
class Solution {
public:
    int strToInt(string str) {
        if (str == "") return 0;
        size_t len = str.size();
        int i = 0;
        long res = 0;
        while (str[i] == ' ') i++;
        if (i == len) return 0;
        bool sign = str[i] == '-' ? false : true;
        if (str[i] == '+' || str[i] == '-') i++;
        while (i < len && isdigit(str[i])) {
            res = res * 10 + str[i] - '0';
            if (sign && res > INT_MAX) return INT_MAX;
            else if (!sign && -res < INT_MIN) return INT_MIN;
            i++;
        }
        return sign ? res : -res;
    }
};

File: hto/67/hto_67.java
class Solution {
    public int strToInt(String str) {
        if (str == null) return 0;
        int n = str.length();
        int i, res = 0;
        boolean is_negative = false;
        for (i = 0; i < n && Character.isWhitespace(str.charAt(i)); ++i);
        if (i == n) return 0;
        if (str.charAt(i) == '-') is_negative = true;
        if (str.charAt(i) == '+' || str.charAt(i) == '-') i++;
        while (i < n && str.charAt(i) >= '0' && str.charAt(i) <= '9') {
            int tmp = str.charAt(i) - '0';  // 减去 '0' 的ASCII码
            if (!is_negative && (res > 214748364 || (res == 214748364 && tmp >= 7))) {
                return 2147483647;
            }
            if (is_negative && (-res < -214748364 || (-res == -214748364 && tmp >= 8))) {
                return -2147483648;
            }
            res = res * 10 + tmp;
            ++i;
        }
        return is_negative ? -res : res;
    }
}

public class leetcode {
    public static void main(String[] args) {
        Solution s = new Solution();
        String str1 = new String("-123");
        String str2 = new String("2147483648");
        System.out.println(s.strToInt(str1) + " " + s.strToInt(str2));
    }
}